src/pkg/packager/composer/list.go

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2021-Present The Zarf Authors
+
+// Package composer contains functions for composing components within Zarf packages.
+package composer
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/defenseunicorns/zarf/src/pkg/layout"
+	"github.com/defenseunicorns/zarf/src/pkg/oci"
+	"github.com/defenseunicorns/zarf/src/pkg/utils"
+	"github.com/defenseunicorns/zarf/src/pkg/utils/helpers"
+	"github.com/defenseunicorns/zarf/src/types"
+)
+
+// Node is a node in the import chain
+type Node struct {
+	types.ZarfComponent
+
+	prev *Node
+	next *Node
+}
+
+// ImportChain is a doubly linked list of components
+type ImportChain struct {
+	head *Node
+	tail *Node
+}
+
+func (ic *ImportChain) append(c types.ZarfComponent) {
+	node := &Node{ZarfComponent: c, prev: nil, next: nil}
+	if ic.head == nil {
+		ic.head = node
+		ic.tail = node
+	} else {
+		p := ic.head
+		for p.next != nil {
+			p = p.next
+		}
+		node.prev = p
+
+		p.next = node
+		ic.tail = node
+	}
+}
+
+// NewImportChain creates a new import chain from a component
+func NewImportChain(head types.ZarfComponent, arch string) (*ImportChain, error) {
+	ic := &ImportChain{}
+
+	ic.append(head)
+
+	history := []string{}
+
+	node := ic.head
+	for node != nil {
+		isLocal := node.Import.Path != "" && node.Import.URL == ""
+		isRemote := node.Import.Path == "" && node.Import.URL != ""
+
+		if !isLocal && !isRemote {
+			// EOL
+			return ic, nil
+		}
+
+		if node.prev != nil && node.prev.Import.URL != "" {
+			return ic, fmt.Errorf("detected malformed import chain, cannot import remote components from remote components")
+		}
+
+		var pkg types.ZarfPackage
+		name := node.Name
+
+		if isLocal {
+			history = append(history, node.Import.Path)
+			paths := append(history, layout.ZarfYAML)
+			if err := utils.ReadYaml(filepath.Join(paths...), &pkg); err != nil {
```

Comment:
This would be nice to do with a helper on the linked list instead of managing it in a separate array.